src/scripts/fabric.js

```diff
@@ -1,6 +1,107 @@
 import * as fabric from "fabric";
 import Engine from "./engine";
 
+// fabric.StaticCanvas.prototype.renderCanvas = function (ctx, objects) {
+//   this.clearContext(this.getContext());
+//   this._renderObjects(ctx, objects);
+// };
+
+const uAffine = (product, curr) =>
+  curr ? fabric.util.multiplyTransformMatrices(curr, product, false) : product;
+
+const u = (product, curr) =>
+  curr ? fabric.util.multiplyTransformMatrices(curr, product, true) : product;
+
+const multiplyTransformMatrixArray = (matrices, is2x2) =>
+  matrices.reduceRight(is2x2 ? u : uAffine, fabric.iMatrix.concat());
+
+class OptimizedRect extends fabric.Rect {
+  m;
+  calcOwnMatrix() {
+    // return [1, 0, 0, 1, this.left, this.top];
+    if (this.m) {
+      return this.m;
+    }
+    const {
+      angle,
+      scaleX,
+      scaleY,
+      flipX,
+      flipY,
+      skewX,
+      skewY,
+      left: x,
+      top: y,
+    } = this;
+
+    const m = multiplyTransformMatrixArray(
+      [
+        // fabric.util.createTranslateMatrix(x, y),
+        angle && fabric.util.createRotateMatrix({ angle }),
+        (scaleX !== 1 || scaleY !== 1 || flipX || flipY) &&
+          fabric.util.createScaleMatrix(
+            flipX ? -scaleX : scaleX,
+            flipY ? -scaleY : scaleY
+          ),
+        skewX && fabric.util.createSkewXMatrix(skewX),
+        skewY && fabric.util.createSkewYMatrix(skewY),
+      ],
+      true
+    );
+
+    m[4] = x;
+    m[5] = y;
+    return (this.m = m);
```

Comment:
is this entirely equivalent? if doing rotate/scale/skew as a 2x2 and then slapping the top/left at the end is the same as starting with the translate as we do today and multiply the ohers as a 2by3, we don't need to swap the order, we can keep the translate as the first step.
Consider also that multiplyTransformMatrixArray should be used in sensitive parts since it adds the inital iMatrix multiplication that is not useful, so you can remove it

Reply:
> if doing rotate/scale/skew as a 2x2 and then slapping the top/left at the end is the same as starting with the translate as we do today and multiply the ohers as a 2by3, we don't need to swap the order, we can keep the translate as the first step.

True, that is why I left it commented out, because if the gain is not huge it is too messy to keep